File: src/engine/executors/http-executor.js
"use strict";

const MAX_RESPONSE_BYTES = 64 * 1024;
const ENV_REF = /^\$\{([A-Za-z_][A-Za-z0-9_]*)\}$/;

function resolveHeaders(headers) {
  const resolved = {};
  for (const [key, value] of Object.entries(headers || {})) {
    const match = typeof value === "string" ? value.match(ENV_REF) : null;
    resolved[key] = match ? process.env[match[1]] || "" : value;
  }
  return resolved;
}

async function readLimitedBody(response) {
  const reader = response.body?.getReader();
  if (!reader) return "";
  const chunks = [];
  let total = 0;
  while (total < MAX_RESPONSE_BYTES) {
    const { done, value } = await reader.read();
    if (done) break;
    const remaining = MAX_RESPONSE_BYTES - total;
    const slice = value.byteLength > remaining ? value.slice(0, remaining) : value;
    chunks.push(slice);
    total += slice.byteLength;
    if (value.byteLength > remaining) break;
  }
  return Buffer.concat(chunks).toString("utf8");
}

async function execute(jobId, scheduledTimeIso, spec) {
  const executionId = `${jobId}:${scheduledTimeIso}`;
  const headers = resolveHeaders(spec.headers);
  if (!headers["User-Agent"]) headers["User-Agent"] = "self-hosted-cron-dispatcher/1.0";
  headers["X-Cron-Execution-Id"] = executionId;

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), spec.timeout_seconds * 1000);
  try {
    const response = await fetch(spec.url, {
      method: spec.method,
      headers,
      body: spec.body || undefined,
      redirect: "manual",
      signal: controller.signal,
    });
    const snippet = await readLimitedBody(response);
    const statusOk = spec.expected_status.includes(response.status);
    const containsOk = spec.validate_contains ? snippet.includes(spec.validate_contains) : true;
    const parts = [`HTTP ${response.status}`];
    if (!statusOk) parts.push("unexpected status");
    if (!containsOk) parts.push("validate_contains not found");
    return { ok: statusOk && containsOk, detail: parts.join("; ") };
  } catch (err) {
    return { ok: false, detail: `request error: ${err.name || "Error"}` };
  } finally {
    clearTimeout(timeout);
  }
}

module.exports = { MAX_RESPONSE_BYTES, resolveHeaders, execute };

File: engine/executors/http_executor.py
"""HTTP job execution with safe defaults: HTTPS-only (enforced upstream by the
job schema), a hard timeout, a bounded/streamed response read, a capped
redirect count, a stable per-occurrence execution-id header, and optional
${ENV_VAR} substitution in header values so a GitHub Secret can be referenced
from a public job file without ever being committed in plaintext.

Never returns or logs response bodies, headers, or raw exception messages that
might embed a URL's query string — only a short, safe summary string.
"""
from __future__ import annotations

import os
import re

import requests

from ..models import HttpSpec

MAX_RESPONSE_BYTES = 64 * 1024
MAX_REDIRECTS = 3

_ENV_REF = re.compile(r"^\$\{([A-Za-z_][A-Za-z0-9_]*)\}$")


def resolve_headers(headers: dict) -> dict:
    """Header values of the exact form ${VAR_NAME} are substituted from the
    process environment at execution time; anything else passes through as-is."""
    resolved = {}
    for key, value in headers.items():
        m = _ENV_REF.match(value) if isinstance(value, str) else None
        resolved[key] = os.environ.get(m.group(1), "") if m else value
    return resolved


def execute(job_id: str, scheduled_time_iso: str, spec: HttpSpec) -> tuple[bool, str]:
    execution_id = f"{job_id}:{scheduled_time_iso}"
    headers = resolve_headers(spec.headers)
    headers.setdefault("User-Agent", "self-hosted-cron-dispatcher/1.0")
    headers["X-Cron-Execution-Id"] = execution_id

    session = requests.Session()
    session.max_redirects = MAX_REDIRECTS
    try:
        resp = session.request(
            spec.method,
            spec.url,
            headers=headers,
            data=spec.body or None,
            timeout=spec.timeout_seconds,
            stream=True,
        )
        try:
            content = bytearray()
            for chunk in resp.iter_content(chunk_size=4096):
                content += chunk
                if len(content) >= MAX_RESPONSE_BYTES:
                    break
            snippet = bytes(content[:MAX_RESPONSE_BYTES]).decode("utf-8", errors="replace")
        finally:
            resp.close()

        status_ok = resp.status_code in spec.expected_status
        contains_ok = (spec.validate_contains in snippet) if spec.validate_contains else True
        ok = status_ok and contains_ok

        parts = [f"HTTP {resp.status_code}"]
        if not status_ok:
            parts.append("unexpected status")
        if not contains_ok:
            parts.append("validate_contains not found")
        return ok, "; ".join(parts)
    except requests.exceptions.RequestException as e:
        return False, f"request error: {type(e).__name__}"
